### DISCUSSION TITLE
LFX Mentorship 2024-01 Pre-test

### DISCUSSION THREAD

**Discussion by hydai** [WasmEdge/WasmEdge#3182] [Show and tell]
### Introduction
This discussion serves as a pre-test guideline for mentorship applications related to proposals #3168, #3169, #3170, and #3172. Completion of this pre-test is mandatory for your application to be considered.

### Pretest details

1. **Framework Execution:** Applicants must demonstrate proficiency in building and executing backend frameworks. This involves working with frameworks such as:
   - [mlx](https://github.com/ml-explore/mlx)
   - [whisper.cpp](https://github.com/ggerganov/whisper.cpp)
   - [intel-extension-for-transformers](https://github.com/intel/intel-extension-for-transformers)
   - [burn](https://github.com/tracel-ai/burn)
   
   You are required to share screenshots and a brief documentation detailing your build and execution process for examples from these frameworks. You can pick any example to demonstrate the execution.

2. **For Proposals #3168, #3169, #3170:**
   - Follow [this guide](https://wasmedge.org/docs/contribute/source/plugin/wasi_nn/#build-wasmedge-with-wasi-nn-llamacpp-backend) to build the `llama.cpp` plugin and execute it with [this chat example](https://github.com/second-state/WasmEdge-WASINN-examples/tree/master/wasmedge-ggml-llama-interactive) or [this API server example](https://github.com/second-state/LlamaEdge/tree/main/api-server).
   - Select any model to run on your device, then share screenshots and a summary of your build process, model selection, and execution results.

3. **For Proposal #3172:**
   - Follow [this guide](https://wasmedge.org/docs/contribute/source/plugin/rusttls) to build and execute the `rustls` plugin.
   - Please choose any example to run the `rustls` plugin on your device, then share screenshots and a summary of your build process and execution results.

4. **Branch Usage:**
   - Due to ongoing updates on the master branch, applicants should use the [hydai/0.13.5_ggml_lts](https://github.com/WasmEdge/WasmEdge/tree/hydai/0.13.5_ggml_lts) branch for this pre-test.

### Submission Format

Upon completing the pre-test, respond to this discussion with the following information:

```
Applied Issue: [Indicate the issue number you are applying for]
GitHub Repo: [Provide a link to your GitHub repository containing the completed tasks from sections 1 & either 2 or 3]
```

For any queries, please post in the designated Q&A thread.

### Deadlines

- **Pretest Deadline:** Submissions must be completed by February 20, 5:00 PM PDT.
- **LFX Applicant Deadline:** LFX applicants must submit by February 13, 5:00 PM PDT.

**Comment by hydai** [comment]
Q&A Thread.

Please ask questions in this thread.
If you believe there is a bug, please create a report on the GitHub Issues.

Thanks.

**Reply by jaydee029** [reply]
Hey @hydai does  using the specified branch make a difference in the execution?, also do we post any error encountered during the process here? or in the respective channel.

**Reply by hydai** [reply]
1. If you did not use the given branch, please indicate which branch you used in the report.
2. If you encounter a bug, please open an issue on GitHub instead of asking here.

**Reply by DhruvSinghiitmandi** [reply]
Hi @hydai 
I was already done with the task 2
Could you please elaborate more on what to report for section 2 as well as section 1

**Reply by hydai** [reply]
> Hi @hydai I was already done with the task 2 Could you please elaborate more on what to report for section 2 as well as section 1

You decide how you present the report. We don't have any limitations on it.

**Reply by RahulSundkar** [reply]
Hi @hydai 
In task 3 for proposal #3172, i'm following the [guide](https://wasmedge.org/docs/contribute/source/plugin/rusttls) that you have specified.  (on WSL(Ubuntu))
When i try to run `cargo install --path .` after building the Rustls plug-in, i get an error saying `error: no packages found with binaries or examples`. 
In the `target/release` folder of the wasmedge_rustls directory, i can see a `libwasmedge_rustls.so` file. 
So am i missing any step in between?

**Reply by trkk28097402** [reply]
Hi @hydai
I see there is a tps performance test, is there any restriction on parameters (n-predict, n-gpu-layers...) to do this test, or they are all editable?
thx!

**Reply by hydai** [reply]
> Hi @hydai In task 3 for proposal #3172, i'm following the [guide](https://wasmedge.org/docs/contribute/source/plugin/rusttls) that you have specified. (on WSL(Ubuntu)) When i try to run `cargo install --path .` after building the Rustls plug-in, i get an error saying `error: no packages found with binaries or examples`. In the `target/release` folder of the wasmedge_rustls directory, i can see a `libwasmedge_rustls.so` file. So am i missing any step in between?

@RahulSundkar Please use the `libwasmedge_rustls.so` in the folder directly. Not sure why this will happen in your environment.

**Reply by hydai** [reply]
> Hi @hydai I see there is a tps performance test, is there any restriction on parameters (n-predict, n-gpu-layers...) to do this test, or they are all editable? thx!

@trkk28097402 
Yes, please refer to this options list: https://github.com/second-state/WasmEdge-WASINN-examples/tree/master/wasmedge-ggml#parameters

Here are some examples:
1. https://github.com/second-state/WasmEdge-WASINN-examples/blob/master/wasmedge-ggml/chatml/src/main.rs#L61-L64
2. https://github.com/second-state/WasmEdge-WASINN-examples/blob/master/wasmedge-ggml/embedding/src/main.rs#L18-L34

**Reply by trkk28097402** [reply]
@hydai 
Thank you very much!
I find that there is a wasmedge-ggml-llama-interactive.wasm file, is it still available?
The remaining wasm files under wasmedge-ggml(4 of them)don't appear to have the same function as interactive one have(which show in the wasmegde's twiiter).

**Reply by hydai** [reply]
@trkk28097402 
You can modify the `src/main.rs`, add any logic you want, and compile it yourself. We just removed these configuration parts to simply the examples. Those configurations are still available.

**Comment by Sidhesh2107** [comment]
hello, I have successfully built WasmEdge with WASI-NN llama.cpp backend. Could you please guide me on how to execute llama.

**Reply by hydai** [reply]
Updated in the section. Please move your question to the Q&A thread, and remove this one.

**Comment by zxc12523** [comment]
Applied Issue: #3168 
GitHub Repo: https://github.com/zxc12523/LFX-Mentorship-2024-01-Pre-test

**Comment by derekwin** [comment]
Applied Issue: #3172 
GitHub Repo: [https://github.com/derekwin/LFX_3172](https://github.com/derekwin/LFX_3172)

**Comment by cr7258** [comment]
Applied Issue: [Integrate whisper.cpp as a new WASI-NN backend](https://github.com/WasmEdge/WasmEdge/issues/3170)
GitHub Repo: https://github.com/cr7258/LFX-Mentorship-WasmEdge-2024-01-Pre-test

**Comment by tannal** [comment]
Applied Issue: https://github.com/WasmEdge/WasmEdge/issues/3170
GitHub Repo: https://github.com/tannal/LFX-Mentorship-WasmEdge-2024-01-Pre-test

**Comment by JackySu** [comment]
Applied Issue: #3172 
Github Repo: https://github.com/JackySu/LFX-Mentorship-2024-01-Pre-test

**Comment by guptaaryan16** [comment]
Applied Issue #3168 
GitHub Repo: https://github.com/guptaaryan16/LFX-WasmEdge-MLX-Pre-Task

**Comment by morristai** [comment]
Applied Issue: https://github.com/WasmEdge/WasmEdge/issues/3172
Github Repo: https://github.com/morristai/WasmEdge-3172-Pre-test

**Comment by souravpy** [comment]
Applied Issue: #3172 
GitHub Repo: https://github.com/souravpy/LFX-Mentorship-WasmEdge-2024-01-Pre-test

**Comment by MaryamTaj** [comment]
Applied issue: #3170 
Github repository: https://github.com/MaryamTaj/LFX-Pre-test
I also wanted to add that completing this pre-test was so much fun! I really enjoyed experimenting with and creating an audio transcription using whisper.cpp - it was so accurate! Thank you so much for this opportunity to learn more about llama.cpp and whisper.cpp :)

**Comment by miketang84** [comment]
Applied Issue: https://github.com/WasmEdge/WasmEdge/issues/3172
GitHub Repo: https://github.com/miketang84/LFX-WasmEdge-3172-Pre-test

**Comment by b1shtream** [comment]
Applied Issue: #3172 
GitHub Repo: [https://github.com/chococandy63/LFX-WasmEdge-Pre-test--3172]( https://github.com/chococandy63/LFX-WasmEdge-Pre-test--3172)

**Comment by XuanDu222** [comment]
Applied Issue: #3170
GitHub Repo: https://github.com/XuanDu222/LFX-Mentorship-2024-01-Pre-test-3182

**Comment by hax0kartik** [comment]
Applied Issue: #3170 
Github Repo: [https://github.com/hax0kartik/LFX-WasmEdge-Pretest](https://github.com/hax0kartik/LFX-WasmEdge-Pretest)

**Comment by hankluo6** [comment]
Applied Issue: https://github.com/WasmEdge/WasmEdge/issues/3170
Github Repo: https://github.com/hankluo6/LFX-Mentorship-Pre-test

**Comment by Sahilgill24** [comment]
Applied issue : #3172 
Github Repo: https://github.com/Sahilgill24/LFX-Mentorship-WasmEdge-2024-01-Pre-test_3172

**Comment by trkk28097402** [comment]
Applied issue: #3169 
Github Repo: https://github.com/trkk28097402/LFX-Mentorship-Pre-test.git

**Comment by grorge123** [comment]
Applied Issue: #3169 
GitHub Repo: https://github.com/grorge123/LFX-Mentorship-WasmEdge-2024-01-Pre-test

**Comment by IRONICBo** [comment]
Applied issue: https://github.com/WasmEdge/WasmEdge/issues/3172
Github Repo: https://github.com/IRONICBo/LFX-WasmEdge-PreTest-3172

Thanks! :)

**Comment by code-sparsh** [comment]
Applied Issue: #3170 
GitHub Repo: https://github.com/code-sparsh/LFX-Pre-Test-WasmEdge

**Comment by RahulSundkar** [comment]
Applied Issue: #3172 
GitHub Repo: [https://github.com/RahulSundkar/LFX_Mentorship-2024_Pre-Test_3172](https://github.com/RahulSundkar/LFX_Mentorship-2024_Pre-Test_3172)

**Comment by cuticle999** [comment]
Applied Issue:[#3170](https://github.com/WasmEdge/WasmEdge/issues/3170)
GitHub Repo:https://github.com/cuticle999/LFX-Mentorship-Pre-test-3170

**Comment by jaydee029** [comment]
Applied under LFX for : #3170
Pretest Repository: https://github.com/jaydee029/LFX-Mentorship-Pre-Test

**Comment by treeleaves30760** [comment]
Applied Issue: #3170 
GitHub Repo: https://github.com/treeleaves30760/LFX_Pre_test

**Comment by rum1887** [comment]
Applied Issue: #3168 
GitHub Repo: https://github.com/rum1887/pretestTaskMLX

**Comment by rum1887** [comment]
Applied Issue: #3170 
GitHub Repo:https://github.com/rum1887/pretestTaskWhisper

**Comment by sandeepB3** [comment]
Applied Issue: #3172 
GitHub Repo: https://github.com/sandeepB3/LFX-Mentorship-2024-PreTest-3172

**Comment by estarriol43** [comment]
Applied Issue: #3168 
GitHub Repo: https://github.com/estarriol43/LFX-MLX-pretest

**Comment by estarriol43** [comment]
Applied Issue: #3169 
GitHub Repo: https://github.com/estarriol43/LFX-itrex-pretest

**Comment by ryan-utopia** [comment]
Applied Issue: #3168 #3170 
GitHub Repo: https://github.com/ryan-utopia/LFX-Mentorship-Pre-test-for-WasmEdge

**Comment by uran0sH** [comment]
Applied Issue: https://github.com/WasmEdge/WasmEdge/issues/3172
GitHub Repo: https://github.com/uran0sH/LFX-Pre-Test-Burn

**Comment by ToSeven** [comment]
Applied Issue: https://github.com/WasmEdge/WasmEdge/issues/3170
GitHub Repo: https://github.com/ToSeven/WasmEdge-2024-01-Pre-test

**Comment by TimmyChiang** [comment]
Applied Issue: #3170
GitHub Repo: https://github.com/TimmyChiang/LFX-Mentorship-2024-01-Pre-test
